File: src/main/java/gg/generations/rarecandy/tools/gui/Source.java
package gg.generations.rarecandy.tools.gui;

public sealed interface Source {
    /** A colour attachment of the G-buffer. */
    record Attachment(int index) implements Source {}

    /** The G-buffer's depth texture. */
    enum Depth implements Source { INSTANCE }

    /** Output of the previous enabled pass, or the G-buffer's albedo if this is the first. */
    enum Previous implements Source { INSTANCE }

    /** A texture from the texture loader, by key. */
    record Named(String key) implements Source {}

    /** Output of an earlier pass that declared {@code writesTo(key)}. */
    record Pass(String key) implements Source {}

    static Source attachment(int index) {
        return new Attachment(index);
    }

    static Source depth() {
        return Depth.INSTANCE;
    }

    static Source previous() {
        return Previous.INSTANCE;
    }

    static Source named(String key) {
        return new Named(key);
    }

    static Source pass(String key) {
        return new Pass(key);
    }
}

File: src/main/java/gg/generations/rarecandy/tools/gui/Source.kt
package gg.generations.rarecandy.tools.gui

interface Source {
    /** A colour attachment of the G-buffer.  */
    @JvmRecord
    data class Attachment(@JvmField val index: Int) : Source

    /** The G-buffer's depth texture.  */
    enum class Depth : Source {
        INSTANCE
    }

    /** Output of the previous enabled pass, or the G-buffer's albedo if this is the first.  */
    enum class Previous : Source { INSTANCE }

    /** A texture from the texture loader, by key.  */
    @JvmRecord data class Named(@JvmField val key: String?) : Source

    /** Output of an earlier pass that declared `writesTo(key)`.  */
    @JvmRecord data class Pass(@JvmField val key: String?) : Source

    companion object {
        @JvmStatic fun attachment(index: Int): Source = Attachment(index)
        @JvmStatic fun depth(): Source = Depth.INSTANCE
        fun previous(): Source = Previous.INSTANCE
        @JvmStatic fun named(key: String?): Source = Named(key)
        fun pass(key: String?): Source = Pass(key)
    }
}